logic/src/commonMain/kotlin/com/wire/kalium/logic/data/client/ClientMapper.kt

```diff
@@ -0,0 +1,85 @@
+package com.wire.kalium.logic.data.client
+
+import com.wire.kalium.logic.data.location.LocationMapper
+import com.wire.kalium.logic.data.prekey.PreyKeyMapper
+import com.wire.kalium.network.api.user.client.Capabilities
+import com.wire.kalium.network.api.user.client.ClientCapabilityDTO
+import com.wire.kalium.network.api.user.client.ClientTypeDTO
+import com.wire.kalium.network.api.user.client.DeviceTypeDTO
+import com.wire.kalium.network.api.user.client.RegisterClientRequest
+import com.wire.kalium.network.api.user.client.RegisterClientResponse
+
+class ClientMapper(
+    private val preyKeyMapper: PreyKeyMapper,
+    private val locationMapper: LocationMapper
+) {
+
+    fun toRegisterClientRequest(param: RegisterClientParam): RegisterClientRequest = RegisterClientRequest(
```

Comment:
would this make sense as an extensions function on RegisterClientParam ?

Reply:
I have mixed feelings about both approaches.

When dealing with mapping of data on Reloaded, we used to test that:
1. Mapper does it's job as a Unit (throw different inputs to it, assert the outputs)
2. Repository called the mapper (or didn't, in case of failures when fetching from data source)
3. Mapper received the correct values from the repository
4. The repository used/returned the mapped values

----

If we want to keep doing that, I see that we have the options:

A. We could mock and stub the extension functions as @tmspzz is suggesting. But there's a compromise: we cannot mock things on iOS. So our mapper-related tests would need to run only on `jvm` and `android` for example.


| Pro                                                                                                  | Con                                                                                     |
|------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------|
| Easier to use in code                                                                                | A bit harder to test, it's possible to mock static/extension calls. Not impossible tho. |
|  | Only testable in JVM and Android                                                        |



B. Just using `ClientMapper` as a concrete class isn't much of an improvement on this regard, as we would need to extend it and create our own fakes in order to test it on all targets. 

| Pro                | Con                              |
|--------------------|----------------------------------|
| Easier to test     | More boiler-plate in code        |
| Clear dependencies | Only testable in JVM and Android |

C. Have a `ClientMapper` interface, and a `ClientMapperImpl`. We can implement fakes of `ClientMapper` during tests for all platforms.

| Pro                     | Con                                                                                    |
|-------------------------|----------------------------------------------------------------------------------------|
| Testable in all targets | Probably the most annoying to test, with more boiler-plate and extensive lines of code |
| Clear dependencies      | A bit more annoying to implement too, as it needs Interface and `Impl` classes         |

D. ? Other ideas ?
----

Between A and B I don't see much of a difference. Neither would allow us to test on all targets. 

If we want to test repositories (as unit, separate from the mappers) and mappers like we used to, C is a necessary evil.

Reply:
If it's customary to have separate *Mapper* classes, don't mind my comments. In Swift I would have probably just made `extensions` in the appropriate files.

Does option B. Take into account that we can subclass it and override whatever we need to in our fakes?

Reply:
Option B was making only one Mapper class, and having it final (which is the default in Kotlin). Mocking libraries allow us to stub them anyway, even if the classes are final. But Mocking libraries = no iOS.

We could, perhaps, do an option D, which would be open Mapper classes instead of final. No need for the `Interface`+`Impl` classes, and we can write our own fakes without mocking libraries.

Makes implementation a bit simpler, but tests stay verbose 🤷🏼‍♂️

Reply:
I think for the sake of simplicity, we should unit test then these mapper classes in isolation inside the `commonTest` folder (1st bullet point). Any other interaction of this mapper class with repositories that we want to test, should live under `AndroidTest` where we can make use of mocking mechanisms.

Reply:
Do you have any existing examples in reloaded of these mapping tests?

I would guess mocking of the mapping classes would be necessary to implements tests of type 2 and 3. Looking at this from the outside, this kind of testing seems like it's testing private logic: correct parameters are forwarded, a mapping is happening or not which would also be covered by tests of type 4.

Besides that I think worth considering what Gonzo suggested. Does all tests need to run on all platforms or can some platform independent logic be tested only by common tests.

Reply:
In my opinion A, B are no go since testing on ios /js is necessary.
C can be a pain in the ass but i don't see it as an issue.
we can mock intefaces in commonTest using something like [this](https://github.com/mockative/mockative) (not really battle tested or fully featuredbut but still a cool concept).